### PR TITLE
added VersionSubscriber with creation of versions

### PR DESCRIPTION
**Pull request by danrot**
This PR adds a new subscriber, which creates a new version everytime a document is published. In addition to that it keeps documents checked out if necessary.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -14,7 +14,8 @@
         "jackalope/jackalope-doctrine-dbal": "~1.2.3",
         "phpunit/phpunit": "^4.5",
         "monolog/monolog": "^1.1",
-        "phpbench/phpbench": "0.10.*"
+        "phpbench/phpbench": "0.10.*",
+        "symfony/phpunit-bridge": "~2.7"
```
Comment (danrot) [inline composer.json]: @wachterjohannes What version should this dependency be set to now? I need it because of the `ClockMock` class.

**Comment by danrot** [comment]
Tests are failing because I had to reintroduce the phpunit-bridge from Symfony for the `ClockMock` class, and some dependencies are containing deprecations again. Should we add the deprecation helper `weak` option` again?

**Comment by danrot** [comment]
Nevermind, fixed it 🙈 Had just to use the correct slugifier class. However, I am still in favour of not using the Symfony CMF slugifier at all, since it doesn't add any value. If we would just use the `Ferrandini/Urlizer` wrapped with our custom interface we would keep control over these things... But that's probably another discussion and out of scope of this PR anyway.

File: lib/DocumentManager.php [hunk 2]
```diff
@@ -137,6 +139,17 @@ public function removeDraft($document, $locale)
     /**
      * {@inheritdoc}
      */
+    public function restore($document, $locale, $version, array $options = [])
+    {
+        $options = $this->getOptionsResolver(Events::FIND)->resolve($options);
```
Comment (chirimoya) [inline lib/DocumentManager.php]: `Events::RESTORE`

File: lib/Subscriber/Behavior/VersionSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+            $versions = $nodes[$versionInformation['path']]->getPropertyValueWithDefault(static::VERSION_PROPERTY, []);
+
+            if (!array_key_exists($versionInformation['path'], $nodeVersions)) {
+                $nodeVersions[$versionInformation['path']] = $versions;
+            }
+            $nodeVersions[$versionInformation['path']][] = json_encode([
+                'locale' => $versionInformation['locale'],
+                'version' => $version->getName(),
+                'author' => $versionInformation['author'],
+                'authored' => date('c', time()),
```
Comment (chirimoya) [inline lib/Subscriber/Behavior/VersionSubscriber.php]: use only `date('c')`

File: lib/Subscriber/Behavior/VersionSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+        if (!$this->support($event->getDocument())) {
+            $event->stopPropagation();
+
+            return;
+        }
+
+        $contentPropertyPrefix = $this->propertyEncoder->localizedContentName('', $event->getLocale());
+        $systemPropertyPrefix = $this->propertyEncoder->localizedSystemName('', $event->getLocale());
+
+        $node = $event->getNode();
```
Comment (chirimoya) [inline lib/Subscriber/Behavior/VersionSubscriber.php:223]: add comment

File: lib/Subscriber/Behavior/VersionSubscriber.php [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        foreach ($node->getProperties() as $property) {
+            if ($this->isRestoreProperty($property->getName(), $contentPropertyPrefix, $systemPropertyPrefix)) {
+                $property->remove();
+            }
+        }
+
+        $version = $this->versionManager->getVersionHistory($node->getPath())->getVersion($event->getVersion());
+        $frozenNode = $version->getFrozenNode();
+
+        foreach ($frozenNode->getPropertiesValues() as $name => $value) {
```
Comment (chirimoya) [inline lib/Subscriber/Behavior/VersionSubscriber.php:236]: add comment

**Comment by chirimoya** [comment]
created and changed date is missing in the live session

**Pull request merged**